RocketChat/Rocket.Chat.ReactNative
# feat: case insensitive for non-ASCII text on main search

## Proposed changes

The proposed change is to create a new column in the database called "sanitized_fname." This column is for the local database only and will store the "fname || name" variable after slugification.

We will use "fname || name" because fname is the string of the channel, discussion, or user that accepts special characters, but in some cases, such as with the #General channel, there is no fname, only the name. If the UTF-8 slugify option in the Workspace admin is enabled, the name will already come with slugify.

The "limax" library is responsible for slugifying in the backend, but I couldn't make it work in React Native. Therefore, I searched for a library that could slugify characters in Russian, traditional Chinese, Japanese, and also Arabic, and I found the "transliteration" library, as can be seen in the unit tests in utils.test.ts.

Regarding the "localSearchSubscription" function, I added a new comparison related to "sanitized_fname," but I needed to keep the other comparisons. When this new database migration is added, only the "sanitized_fname" column is added, and the slugify is not done for all subscriptions already present in the local db. The slugify of these subscriptions will be done as the user uses the app, for example, when entering a room, receiving a new message, or changing the room name. Therefore, over time, all subscriptions will have the sanitized_fname column written, except for very old rooms that the user has never opened or changed. Another case to have all sanitized_fname is to perform a clear cache, which would also slugify "fname || name" when rooms are created in the local db.

Why all of this?
Well, it all started with a community issue reporting that local search was case-sensitive for Cyrillic characters on iOS, while on Android, this search was case-insensitive. So, I researched if it was possible to change WatermelonDB or FMDB (the objective C library used by WatermelonDB for iOS), but the result was that it would be much more complex. In WatermelonDB documentation, there's a note saying: "For instance, while the comparison is case-insensitive, SQLite cannot by default compare non-ASCII characters case-insensitively (unless you install ICU extension)". However, I could not find a step-by-step guide to install this extension, and as it's only for iOS, we would have to change a lot of native code and keep maintaining it every time we updated the library. So, I abandoned this idea and went for the slugify approach.

Although there is only one open issue from a while ago, this fix is very important. Otherwise, users who have rooms with non-Latin characters will have to search through the list or remember upper and lower case letters.

## Issue(s)	

Closes https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/3276

## How to test or reproduce

- Create a channel `ПРОВЕРКА`
- On `RoomListView` click on the search icon
- Write `Пр`
- Should return the channel created

## Screenshots
<table>
<thead>
<tr>
<th>Before</th>
<th>After</th>
</thead>
</tr>
<body>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/47038980/127541738-90471066-66aa-4af5-9b1f-4c7293ea3a33.png"  height="450"/>
</td>
<td>
<img src="https://user-images.githubusercontent.com/47038980/127541804-e6915330-cbd9-4c14-b3fb-b231a0af9c68.png" height="450" />
</td>
</tr>
</tbody>
</table>

## Types of changes




## Checklist



## Further comments

Closes https://github.com/RocketChat/Rocket.Chat.ReactNative/issues/3276
[TC-394]

[TC-394]: https://rocketchat.atlassian.net/browse/TC-394?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
